src/main/kotlin/kweb/html/ElementReader.kt

```diff
@@ -11,7 +12,9 @@ open class ElementReader(protected val receiver: WebBrowser, internal val elemen
     constructor(element: Element) : this(element.browser, element.id)
 
     init {
-        require(!receiver.kweb.isCatchingOutbound()) {
+        //TODO I'm not sure if we want to allow reading the DOM during a render or non immediate event
```

Comment:
Hmm, this should throw an error during an immediate event, but this is a much more complicated question when it comes to a `render` or a non-immediate event handler.

I think we might need to backtrack a little and make `Kweb.callJsWithCallback()` throw an error if called during an `IMMEDIATE_EVENT`, but bypass the catcher during a `BATCH`. This would probably require adding a `bypassCatcher` parameter to `callJs()` which defaults to `false`.

Reply:
This is definitely complex. I was ready to completely agree with you and implement your solution here. But, first one question.
Are we sure it's safe to have `IMMEDIATE_EVENT`'s bypass the catcher? Could it be possible that an `immediateEvent` get called on an element that gets created or modified before the immediateEvent is supposed to run? Bypassing the messageCatcher could lead to something happening in the wrong order, or just outright failing if an element doesn't exist yet.

Reply:
You make a good point about out-of-order execution, although it won't happen during an `IMMEDIATE_EVENT` because that should throw an error, only during a `BATCH`.

However this does mean that the `callJsWithCallback()` will be sent to the client before previous instructions that are batched, and this could cause problems.

Solving this may require some changes to how `catchOutbound()` works. I think the solution is for a call to `callJsWithCallback()` should trigger a "flush" of the caught `FunctionCall`s up to and including anything added by `callJsWithCallback()`, it should then resume catching as before. This will ensure that instructions are sent to the client in-order as before, but also that `callJsWithCallback()` calls get sent to the client without delay.

This will require a rethink of how `catchOutbound()` works, perhaps changing it to something like:

```kotlin
    fun catchOutbound(flushOutbound : (List<FunctionCall>) -> Boolean, f: () -> Unit): List<FunctionCall> {
```
By supplying a `flushOutbound` function which can be used multiple times during the execution of `f` I think it will give us this flexibilty.

Reply:
Something I meant to put in my previous comment, is also, what problem is this a solution to? Why does a non immediate event need to immediately handle the callback? What's wrong with the callback running at the end of the render() or event code block, as long as everything runs in the correct order on the client?

I just don't see what we gain from this added complexity. Also, this would break our batching, because if someone adds 5 elements to a list, and runs a callback on the 3rd one, the client will add the first 3, and flicker before adding the last 2.

Reply:
A scenario I'm thinking of would be where we retrieve something from the DOM, perhaps the value of an `<input>`, and then attempt to use it later within the same batch. I think this would create a deadlock because the batch would stop and wait but the Function to retrieve the DOM data wouldn't get to execute.

But perhaps the simpler solution, which would be to disallow DOM retrievals within batches - is preferable, I'm trying to think through whether this would disallow anything important. There are probably other issues with anything that suspends a batched block (like an `.await()`), because the part after the suspension may not be the same thread, which would screw up the threadlocal stuff.

I need to ponder it a bit further.